### PR TITLE
linkd

### PR DESCRIPTION
**Pull request by xla**
First iterative implementation of [RFC 0696][1] focusing on driving the core
peer protocol and establishing the harness for configuration of a running node.
Rite of passage for it was the replacement of the ephemeral peer in the e2e
networkss, as that required a well behaved peers with a set of knobs exposed.

Structurally the majority implementation lives in a library crate `linkd-lib` -
probably the most horrendous crate name in this repo to date - with a newly
created workspace under `bins` and a `linkd` crate in wrapping it lightly in a
`main`. That strategy allows for a checked-in `Cargo.lock` for binaries
associated with link project. Beyond that the code is rather Mario-esque as it
is primarily plumbing of code that either existed in other crates backing
binaries or newish code doing the same thing for core pieces.

There are still large pieces missing, which a tracked in [#722][2] and will be
piled on-top as patches to keep this already big delta focused. The declared
initial goal is to get `linkd` into a state where it can run as bootstrap/seed
node.

[1]: https://github.com/radicle-dev/radicle-link/blob/master/docs/rfc/0696-p2p-node.adoc
[2]: https://github.com/radicle-dev/radicle-link/issues/722

### CODE REVIEW

**Review by FintanH — COMMENTED** [review]
Just some small things so far, but it's looking like it's coming together :)

File: librad/src/profile.rs
Comment (xla) [reply librad/src/profile.rs:55]: Structopt requires a `Display` impl.

Comment (FintanH) [reply librad/src/profile.rs:55]: Sorry, I wasn't clear enough. I'm suggesting the following:
```rust
impl fmt::Display for RadHome {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}
```

So we're just setting the `Display` implementation to be the same as the `Debug` one, which is almost what you're doing here but with slight differences.

File: linkd/src/cfg/seed.rs
Comment (FintanH) [inline linkd/src/cfg/seed.rs:36]:
```suggestion
    async fn try_from_bootstrap(bootstrap: &Bootstrap) -> Result<Self, Error> {
```

For idiotmaticness

File: linkd/src/lib.rs
Comment (FintanH) [inline linkd/src/lib.rs:8]: bikeshed: we've used `config` or `Config` everywhere else, so I'd vote we stay consistent on the naming.

File: linkd/src/args.rs
Comment (FintanH) [reply linkd/src/args.rs:47]: Maybe not because these resolve to `SocketAddr`, right?

File: librad/src/profile.rs
Comment (xla) [reply librad/src/profile.rs:55]: a03bd32

File: bins/Cargo.toml
Comment (xla) [reply bins/Cargo.toml]: 44bf404

File: linkd/src/args.rs
Comment (xla) [reply linkd/src/args.rs]: This takes precedence.

File: linkd/src/args.rs
Comment (xla) [reply linkd/src/args.rs]: Don't think URL will serve us well, we don't use much of it otherwise.

File: linkd/src/lib.rs
Comment (xla) [reply linkd/src/lib.rs]: Can you point me to an approved RFC where we agreed on this? :bike: :derelict_house:

File: linkd/src/cfg/seed.rs
Comment (xla) [reply linkd/src/cfg/seed.rs]: It would be more idiomatic if the advent of async wouldn't been incompatible with a majority of language idioms...

485bfbd

File: librad/src/profile.rs
Comment (FintanH) [reply librad/src/profile.rs:55]: :heart:

File: linkd/src/args.rs
Comment (FintanH) [reply linkd/src/args.rs]: But if it's not specified then it defaults to `RadHome::ProjectDirs`, so I could do:
```
$ RAD_HOME=/tmp/rad-home linkd --ssh-agent
```
and then it would end up using the project dirs

File: linkd/src/args.rs
Comment (FintanH) [reply linkd/src/args.rs]: Yup, fair :)

File: linkd/src/cfg/seed.rs
Comment (FintanH) [reply linkd/src/cfg/seed.rs]: That's why it's idiotmatic =]

File: linkd/src/lib.rs
Comment (FintanH) [reply linkd/src/lib.rs]: touché

File: linkd/src/args.rs
Comment (xla) [reply linkd/src/args.rs]: We could employ structopts magic of reading env variables. I wonder tho, will this be hard to communicate what arg will have what value - the matrix of combinations is expanding. Furthermore the daemon is unlikely be run just from a users tty. Where and when is it usual to set the `RAD_HOME`?

Comment (FintanH) [reply linkd/src/args.rs]: > Where and when is it usual to set the RAD_HOME?

I would say that it should be used for testing or someone who _really_ wants their home to be somewhere else. So I think this parameter should just be removed. 

For its use here[[0]], I would recommend this usage[[1]].

[0]: https://github.com/radicle-dev/radicle-link/pull/779/files#diff-690b715b0168b0abda74382805e6b80dd3e8ffed9bd71b7b7739d281361ea959R125-R131
[1]: https://github.com/radicle-dev/radicle-link/blob/master/rad-profile/src/lib.rs#L72

Comment (xla) [reply linkd/src/args.rs]: I don't think I agree, if we remove this argument then in order to fully configure the daemon you are forced to employ a mi of env vars and cli options.

Comment (xla) [reply linkd/src/args.rs]: > For its use here[[0](https://github.com/radicle-dev/radicle-link/pull/779/files#diff-690b715b0168b0abda74382805e6b80dd3e8ffed9bd71b7b7739d281361ea959R125-R131)], I would recommend this usage[[1](https://github.com/radicle-dev/radicle-link/blob/master/rad-profile/src/lib.rs#L72)].

Sorry I was looking for places where the env var `RAD_HOME` is set.

Comment (FintanH) [reply linkd/src/args.rs]: > I don't think I agree, if we remove this argument then in order to fully configure the daemon you are forced to employ a mix of env vars and cli options.

I think in this particular case it's not going to be set that often. So you won't realistically be mixing env and cli options. 

> Sorry I was looking for places where the env var RAD_HOME is set.

Upstream's sandbox environment :shrug: It doesn't tend to be set for actual use.

Comment (xla) [reply linkd/src/args.rs]: > Upstream's sandbox environment shrug It doesn't tend to be set for actual use.

Reinforces my impression to try to weave in these env vars and always expect explicit configuration through the cli.

Comment (FintanH) [reply linkd/src/args.rs]: Then you need to find a way for this CLI parameter to play well with the env var

Comment (xla) [reply linkd/src/args.rs]: > Then you need to find a way for this CLI parameter to play well with the env var

If it's never set, why bother keeping it the env var around?

Comment (FintanH) [reply linkd/src/args.rs]: It doesn't mean it _can't_ be set. Like I said above, the following command would have undesirable effects:
```
$ RAD_HOME=/tmp/rad-home linkd --ssh-agent
```

Comment (xla) [reply linkd/src/args.rs]: 1b60712

File: linkd-lib/src/cfg.rs
Comment (alexjg) [inline linkd-lib/src/cfg.rs:50]: nit: should be `Io`

File: linkd-lib/src/socket_activation/unix.rs
Comment (alexjg) [inline linkd-lib/src/socket_activation/unix.rs:35]: What would be the purpose of using multiple file descriptors?

Comment (xla) [reply linkd-lib/src/socket_activation/unix.rs:35]: Systemd unit files support the declaration of N named sockets, these are then used by the binary at own discretion. Use-cases include but not limited to offering multiple APIs or offer telemetry on an isolated socket.

> sd_listen_fds_with_names() is like sd_listen_fds(), but optionally also returns an array of strings with identification names for the passed file descriptors

From https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html

Comment (alexjg) [reply linkd-lib/src/socket_activation/unix.rs:35]: Gotcha, thanks.

File: linkd-lib/Cargo.toml
Comment (kim) [inline linkd-lib/Cargo.toml:23]: This does not support Windows, as it uses a "pipe" which actually is a tokio `UnixStream` for signal delivery. It saves a thread, yes, but I'm not sure this is overall cheaper.

File: linkd-lib/Cargo.toml
Comment (kim) [inline linkd-lib/Cargo.toml:27]: Do you need `signal`?

File: linkd-lib/src/args.rs
Comment (kim) [inline linkd-lib/src/args.rs:53]: IIUC, setting this to `true` will render `rad_home` a no-op. Perhaps fold into one argument?

File: linkd-lib/src/args.rs
Comment (kim) [inline linkd-lib/src/args.rs:192]: I was wondering: "whose memory?"

Maybe this should be called `Ephemeral`, `AdHoc`, `Generate`, or somesuch?

File: linkd-lib/src/socket_activation/macos.rs
Comment (kim) [inline linkd-lib/src/socket_activation/macos.rs:11]: :shrimp:

File: linkd-lib/src/logging.rs
Comment (FintanH) [inline linkd-lib/src/logging.rs:1]: I was thinking, logging might be useful to the `rad` CLI as well, so wondering if we'd be interested in putting this in `rad-clib`?

File: Cargo.toml
Comment (cloudhead) [reply Cargo.toml:15]: `link-daemon`?

File: linkd-lib/src/args.rs
Comment (cloudhead) [inline linkd-lib/src/args.rs:198]: I'd think `File` is a more sensible default, otherwise we block on stdin and it's confusing

Comment (cloudhead) [reply linkd-lib/src/args.rs:198]: In fact not sure why you'd want stdin if you have file

File: linkd-lib/src/cfg.rs
Comment (cloudhead) [inline linkd-lib/src/cfg.rs:41]: `([0,0,0,0], 0).into()`?

File: linkd-lib/src/cfg/seed.rs
Comment (cloudhead) [inline linkd-lib/src/cfg/seed.rs:37]: Why not `collect` so we can load multiple seeds here?

File: linkd-lib/src/logging.rs
Comment (cloudhead) [inline linkd-lib/src/logging.rs:27]: Maybe:
```rust
                .with_env_filter(
                    EnvFilter::try_from_default_env()
                        .unwrap_or_else(|_| tracing_subscriber::EnvFilter::new("debug")),
                )
```

File: linkd-lib/src/protocol.rs
Comment (cloudhead) [inline linkd-lib/src/protocol.rs:35]: I hate that this is a thing

Comment (cloudhead) [reply linkd-lib/src/protocol.rs:35]: :mage_man: :sparkles:

File: Cargo.toml
Comment (kim) [reply Cargo.toml:15]: I liked the `blink`

File: linkd-lib/src/cfg.rs
Comment (kim) [reply linkd-lib/src/cfg.rs:41]: That’s not a good default, either. I’d rather mandate that the bind address must be specified explicitly, and provide shortcuts such as “any”, “localhost”

File: Cargo.toml
Comment (xla) [reply Cargo.toml:15]: In my understanding `linkd-lib` is adhering to our non-RFCed standard of crate naming, `linkd` being the prefix of the binary that consumes this crate (same as the `rad` crates for the cli).

For comedic purpose and artificial increase of barrier to entry for contributors I'm not opposed to naming it `dlink`.

File: linkd-lib/src/socket_activation/macos.rs
Comment (xla) [reply linkd-lib/src/socket_activation/macos.rs:11]: :apple:

File: linkd-lib/Cargo.toml
Comment (xla) [reply linkd-lib/Cargo.toml:21]: Not your kind of nix, boy!

File: linkd-lib/src/node.rs
Comment (xla) [reply linkd-lib/src/node.rs:52]: :bacon: :fried_egg:

File: linkd-lib/Cargo.toml
Comment (xla) [reply linkd-lib/Cargo.toml:23]: Was mostly optimising for ergonomics as this fits better with the style of spawning a task per subroutine. While the failing windows tests haunted me in my dreams the last days I concluded it might be worthwhile porting the parked thread with atomics approach we have in other places over and give the impression that this project works on Windows after all. Maybe then @FintanH stops calling me out on it.

File: linkd-lib/Cargo.toml
Comment (xla) [reply linkd-lib/Cargo.toml:27]: Maybe not, or maybe I drop the other signal related code in favour of this. What could go wrong, tokio is production ready software, right?

File: linkd-lib/src/args.rs
Comment (xla) [reply linkd-lib/src/args.rs:6]: With our rapid offboarding of this walled garden I'm uncertain where to track follow-up work. Will go over the TODOs and see if they are out-of-sync with the #722.

File: linkd-lib/src/args.rs
Comment (xla) [reply linkd-lib/src/args.rs:251]: You are right, I haven't found a way to enforce a heuristic a la: only expect X when Y is set.

File: linkd-lib/src/metrics.rs
Comment (xla) [reply linkd-lib/src/metrics.rs:6]: Besides this code path it wouldn't guard any dependencies as this doesn't have any besides `std`. In my understanding that is usually why crate consumers would want feature flags to not be exposed to certain sub-trees in the cargo tree. Not sure where we generally stand on the matter as to what should and needs to be enabled with flags.

File: linkd-lib/src/logging.rs
Comment (xla) [reply linkd-lib/src/logging.rs:1]: Generally yes, but we should build a proper :diamond_shape_with_a_dot_inside: and move the shared code (signer things, logging, etc.) into its own crate, otherwise linkd would depend on rad, which strikes me as out.

File: linkd-lib/src/args.rs
Comment (xla) [reply linkd-lib/src/args.rs:198]: Sometime in the past, someone in an outburst of productivity produced overnight a critical piece of infrastructure. To aid the network in bootstrapping and bring replication to the plebs, born was the SEED. As the creator believed in composability and drank from the firehose called the UNIX pipe, they made it so that you can pipe the most private of private things - the secret key -  into the SEEDs stdin. It was glorious, but many an operator lost their sleep over the breaking systemd services, a lot of bash was written in those days (dark times indeed). And because of that legacy we have support for STDIN.

In fact, I thought I never say that, it has proven handy for configuration of the whale containers. There you often share the same image for different peers, so you can't really store a single key on disk as they would share it. Changing the default to file is not an outrageous thought.

File: linkd-lib/src/cfg/seed.rs
Comment (xla) [reply linkd-lib/src/cfg/seed.rs:56]: To me that seems like a configuration error and should be surfaced asap. Main reason being is that if we just bubble it up as log line, it will become the typical noise in operational logs and operators will not be acting on it.

File: linkd-lib/src/protocol.rs
Comment (xla) [reply linkd-lib/src/protocol.rs:35]: Unfortunately fixing async Rusts Pinning abstraction is out-of-scope for this PR. :<

File: linkd-lib/src/cfg/seed.rs
Comment (xla) [reply linkd-lib/src/cfg/seed.rs:37]: Because they would share the same `peer_id` and that is a violation of the protocol. But you are absolutely right that we should collect all addresses and associate them with the same peer, as the discovery expects a stream of `(PeerId, Vec<SocketAddr>)`.

File: linkd-lib/src/args.rs
Comment (xla) [reply linkd-lib/src/args.rs]: ac88165

File: linkd-lib/src/args.rs
Comment (xla) [reply linkd-lib/src/args.rs]: e1f8531

File: linkd-lib/src/cfg/seed.rs
Comment (xla) [reply linkd-lib/src/cfg/seed.rs]: 9d993d9

File: linkd-lib/src/logging.rs
Comment (xla) [reply linkd-lib/src/logging.rs]: 7cf3e34

File: linkd-lib/src/cfg.rs
Comment (xla) [reply linkd-lib/src/cfg.rs:45]: bd8ee1b

File: linkd-lib/Cargo.toml
Comment (kim) [reply linkd-lib/Cargo.toml:27]: tokio only knows how to handle `CTRL-C` on windows (there's a dedicated API for that). Maybe that's enough, idk, how does one run services there anyway?

File: linkd-lib/src/metrics.rs
Comment (FintanH) [reply linkd-lib/src/metrics.rs:6]: The `std` point is good :) Happy to not stick it behind a flag :ok_hand:

File: linkd-lib/src/args.rs
Comment (FintanH) [reply linkd-lib/src/args.rs:249]: Ya, I meant the way one would usually do this is by packing it as an enum variant, right?
```rust
enum Provider {
   Graphite { address: String }
}
```
but that doesn't fly if we want flattened args, so I guess we're between a rock and a hard place :) I wonder would it make sense to have the `graphite_addr` as `Option` as well?

**Review by FintanH — APPROVED** [review]
Nice work =] Looking forward to the new age of daemons

File: Cargo.toml
Comment (kim) [reply Cargo.toml:15]: > the prefix of the binary that consumes this crate

Interesting, this interpretation didn't occur to me. I'm thinking of `rad` as an anagram of `cli`, where `cli` is a bunch of commandline executables, not necessarily that they must be invoked by typing `rad` first (the fact that we said it will be one executable for starters is an implementation detail if you will). I'm also thinking of these prefixes as a navigational aid in a monorepo structure, while I discarded the idea to enforce complete separation -- I would expect that there will always be things which seem worthwhile to put in their own crate, so they can be depended on in multiple places.

So tl;dr I feel `linkd-` as a top-level name is too specific, because we may have other daemons. Maybe we can find a different prefix conveying daemon-ness? It could be `daemon-`, although speling...

Comment (FintanH) [reply Cargo.toml:15]: > different prefix conveying daemon-ness?

How about `matt`? 

image

File: linkd-lib/Cargo.toml
Comment (xla) [reply linkd-lib/Cargo.toml:25]: I concluded until we have more insights into how link is operate on gaming rigs break and ctrl-c is what can be expected.

c792721

File: linkd-lib/src/cfg/seed.rs
Comment (cloudhead) [reply linkd-lib/src/cfg/seed.rs]: well, DNS can fail temporarily though, but yeah up to you!

File: linkd-lib/src/args.rs
Comment (cloudhead) [reply linkd-lib/src/args.rs]: Epic

File: Cargo.toml
Comment (xla) [reply Cargo.toml]: 3d4e38c

File: linkd-lib/src/args.rs
Comment (xla) [reply linkd-lib/src/args.rs]: Might be I'm holding the Structops wrong but that would conflict with the need to default to the local graphite instance if graphite is set as a metric provider.

Comment (FintanH) [reply linkd-lib/src/args.rs]: My thinking is that if `metrics-provider` is set to `"graphite"` then you'll end up with `Some(MetricsProvider::Graphite)` and then the address will be `Some("localhost:8080")` (or whatever the port is). And on the flip-side, if `metrics-provider` is `None` then `graphite_addr` is also, rightfully, `None`. 

That being said this is not a big deal, and should not be a blocker :) Just musing about what's the best way to represent the data

Comment (xla) [reply linkd-lib/src/args.rs]: I don't follow how you'd express that with the structopt primitives.

Comment (FintanH) [reply linkd-lib/src/args.rs]: My suggestion is:
```rust
#[derive(Debug, Eq, PartialEq, StructOpt)]
pub struct MetricsArgs {
    /// Provider for metrics collection.
    #[structopt(long = "metrics-provider", name = "metrics-provider")]
    pub provider: Option<MetricsProvider>,

    /// Address of the graphite collector to send stats to.
    #[structopt(
        long,
        required_if("metrics-provider", "graphite")
    )]
    pub graphite_addr: Option<String>,
}

/* then the conversion here */
        let metrics = match args.metrics.provider {
            Some(args::MetricsProvider::Graphite) => Some(Metrics::Graphite(
                args.metrics
                    .graphite_addr
                    .unwrap_or("localhost:2003")
                    .to_socket_addrs()?
                    .next()
                    .unwrap(),
            )),
            None => None,
        };
```

Which is a minor improvement if any. It's just unfortunate that we can't express this better in `structopt`.

File: linkd-lib/src/cfg/seed.rs
Comment (FintanH) [reply linkd-lib/src/cfg/seed.rs]: > well, DNS can fail temporarily though, but yeah up to you!

Perhaps a retry mechanism can be added in a future iteration?

**Review by FintanH — DISMISSED** [review]

**Review by kim — APPROVED** [review]

**Comment by kim** [comment]
I would, however, appreciate a slightly more condensed commit history.

**Comment by xla** [comment]
> I would, however, appreciate a slightly more condensed commit history.

I always squash before merge but leave that to the latest point to account for commits addressing feedback.

**Pull request merged**
